add warmup lr_scheduler

solin319 opened the pull request:
Add warmup lr_scheduler, it can overcome optimization challenges early in large batch-size.
This method was used in "Accurate, Large Minibatch SGD: Training ImageNet in 1 Hour".
https://arxiv.org/abs/1706.02677

zhreshold commented:
Warmup behavior is nice.
Do you mind adding a combined warmup and multi step scheduler which will be more useful IMO?

solin319 commented:
Yes, I will combine warmup and multi step scheduler in a single class.

solin319 commented:
I have combined warmup and multi step scheduler in a single class.
It can be used as  
“lr_scheduler = mx.lr_scheduler.GradualWarmupWithMultiFactorScheduler(warmup_step=10, begin_lr=0.002, stop_lr=0.1, multifactor_step=[15,18,20], factor=0.1)”

zhreshold commented on python/mxnet/lr_scheduler.py:
```diff
@@ -136,3 +136,80 @@ def __call__(self, num_update):
+        super(GradualWarmupWithMultiFactorScheduler, self).__init__()
+        if stop_lr <= begin_lr:
+             raise ValueError("stop_lr must larger than begin_lr")
+        assert isinstance(multifactor_step, list) and len(multifactor_step) >= 1
```
default multifactor_step == int will break, add cast to list first

zhreshold commented on python/mxnet/lr_scheduler.py:
```diff
@@ -136,3 +136,80 @@ def __call__(self, num_update):
+                raise ValueError("Schedule step must be an increasing integer list")
+            if _step < 1:
+                raise ValueError("Schedule step must be greater or equal than 1 round")
+        if multifactor_step[0] <= warmup_step:
```
default step 100 will break here too

piiswrong commented:
Please add test.

Can this be added as an additional argument of MultiFactorScheduler?

solin319 commented:
I have added  additional argument in MultiFactorScheduler class.
Warmup can be used when 'warmup_step' > 1.

I added a test file in test/commom/test_lr_sceduler.py.
Following is a test result
```
python test_lr_scheduler.py --lr-factor=0.1 --warmup_step=5 --lr-steps='8,10'

INFO:root:step 1 lr = 0.020000
INFO:root:step 2 lr = 0.040000
INFO:root:step 3 lr = 0.060000
INFO:root:step 4 lr = 0.080000
INFO:root:Update[5]: now learning rate arrived at 1.00000e-01, will not warm up in the future
INFO:root:step 5 lr = 0.100000
INFO:root:step 6 lr = 0.100000
INFO:root:step 7 lr = 0.100000
INFO:root:step 8 lr = 0.100000
INFO:root:Update[9]: Change learning rate to 1.00000e-02
INFO:root:step 9 lr = 0.010000
INFO:root:step 10 lr = 0.010000
INFO:root:Update[11]: Change learning rate to 1.00000e-03
INFO:root:step 11 lr = 0.001000
INFO:root:step 12 lr = 0.001000
INFO:root:step 13 lr = 0.001000
INFO:root:step 14 lr = 0.001000

```

solin319 commented:
I have moved test_lr_scheduler in tests/python/unittest/test_lr_scheduler.py.
I  have removed argparse and added some test cases in the test file.
@zhreshold

zhreshold commented on tests/python/unittest/test_lr_scheduler.py:
```diff
@@ -0,0 +1,57 @@
+        assert mx.test_utils.almost_equal(res_lr[steps[i]], lr * pow(lr_factor, i + 1), 1e-10)    
+
+if __name__ == "__main__":
+    #Legal input
```
Wrap all these tests in a function like test_multi_lr_scheduler(), and use 
```
if __name__ == '__main__':
    import nose
    nose.runmodule()

# or maybe without nose
if __name__ == '__main__':
    test_multi_step_lr_scheduler()
```
to trigger all tests, we can have more unit test for lr_schedulers in the future.

zhreshold commented on tests/python/unittest/test_lr_scheduler.py:
```diff
@@ -0,0 +1,57 @@
+    test_lr_sceduler(lr = 0.2, steps = [8,12], lr_factor = 0.1, warmup_step = 0, warmup_lr = 0.1)
+    test_lr_sceduler(lr = 0.02, steps = [8,12], lr_factor = 0.1, warmup_step = 1, warmup_lr = 0.1)
+    test_lr_sceduler(lr = 0.02, steps = [8,12], lr_factor = 0.3, warmup_step = 5, warmup_lr = 0.1)
+    test_lr_sceduler(lr = 0.002, steps = [8,12], lr_factor = 0.1, warmup_step = 7, warmup_lr = 0.1)
```
add a default test like: test_lr_sceduler(steps=[100, 200]) to test the default parameters if user don't specify anything.

solin319 commented:
I have used 'nose' to test and add default test 'multi_lr_sceduler(lr = 0.02, steps=[100, 200])'.
@zhreshold

zhreshold commented:
please `make pylint` at root, fix potential error and wait for CI to finish.

solin319 commented:
Add pylint and license as "test_operator.py"

zhreshold commented:
You are all set. Can you rebase master and push again?, the checks failed with other problems.

solin319 commented:
move to #8107
